test/quantization/test_workflow_module.py

```diff
@@ -117,27 +117,29 @@ def _permute_to_axis_zero(X, axis):
 
 # Reference method for fake quantize
 def _fake_quantize_per_channel_affine_reference(X, per_channel_scale, per_channel_zero_point, axis, quant_min, quant_max):
-    X, permute_axis_list = _permute_to_axis_zero(X, axis)
+    dtype = X.dtype
+    X, permute_axis_list = _permute_to_axis_zero(X.to(torch.float32), axis)
```

Comment:
do I understand correctly that from L121 to L128, `X` has dtype of `torch.float`?  If yes, should that be changed?  I would have expected the dtype to be consistent with dtype under test throughout the test case.

Reply:
the issue is the reference function doesn't support half precision floats. So I have to convert to float32. 

This is also what we did with per_tensor: https://github.com/pytorch/pytorch/blob/a90a3acbeec7deebecf7baf53c8f0fdf7b37e5ae/test/quantization/test_workflow_module.py#L57

is there a different way to implement the reference function that I should be using?

Reply:
which specific ops do not support fp32?  Can we just leave those ops in fp32, and do the rest in fp16?

Reply:
rounding is the main issue, I don't see how it changes anything though, if I added .to(torch.float) within the calculation for res, its funcitonally the same, right?

Reply:
IMO it's not worth it to implement fp16 support for torch.round just to make this test pass, so this reference function isn't really the correct answer.

I do think it's worth minimizing the differences between what this function is doing comparing to what it could do if all ops supported fp16.  It could be something like

```
# old
 res[i] = (torch.clamp(torch.round(X[i] * (1.0 / per_channel_scale[i]) + per_channel_zero_point[i]), quant_min, quant_max) - per_channel_zero_point[i]) * per_channel_scale[i]

# new
 scaled_correct_dtype = X[i] * (1.0 / per_channel_scale[i]) + per_channel_zero_point[i]
 scaled_fp32 = scaled_correct_dtype.float()
 rounded = torch.round(scaled_fp32)
 rounded_correct_dtype = rounded.to(correct_dtype)
 res[i] = (torch.clamp(rounded_correct_dtype, quant_min, quant_max) - per_channel_zero_point[i]) * per_channel_scale[i]
```

Reply:
note: clamp is also not supported for fp16

Reply:
ok, then how about

```
# torch.clamp is not supported in fp16
def _clamp_in_float(x, min_val, max_val):
  old_dtype = x.dtype
  x = x.to(torch.float)
  x = torch.clamp(x, min_val, max_val)
  x = x.to(old_dtype)
  return x

# torch.round is not supported in fp16
def _round_in_float(x, min_val, max_val):
  old_dtype = x.dtype
  x = x.to(torch.float)
  x = torch.round(x)
  x = x.to(old_dtype)
  return x

res[i] = (_clamp_in_float(_round_in_float(X[i] * (1.0 / per_channel_scale[i]) +
                  per_channel_zero_point[i]), quant_min, quant_max) - per_channel_zero_point[i]) * per_channel_scale[i]
```

Reply:
@HDCharles found that this is due to the scale being in fp32.  The current behavior is expected then.  Would be great to add a comment to the code for future readers.